### ISSUE TITLE
wrong order in cypher query

### ISSUE THREAD

**Issue by matiasmorant** [neo4j/neo4j#12760]
I discovered a cypher query that returns results in wrong order

Neo4j Version: 4.2.3
Operating System: Windows 10
API: Cypher

##  Steps to reproduce

1- create a graph

`CREATE ({num:2})-[:r1]->({num:0})-[:r2]->({num:1})`

2- execute the following query

```
MATCH (x) WITH x ORDER BY x.num WITH x CALL {
    WITH x CALL {
        WITH x MATCH (x)-[y :r1]->() RETURN y
        UNION
        WITH x MATCH (x)-[y :r2]->() RETURN y
    } WITH y RETURN collect(y) as y
} WITH x RETURN x.num
```
(the CALL matches all edges but shouldn't change the result. needed to reproduce the bug)

##  Expected behavior

the query should return
```
╒═══════╕
│"x.num"│
╞═══════╡
│0      │
├───────┤
│1      │
├───────┤
│2      │
└───────┘
```
## Actual behavior

query returns

```
╒═══════╕
│"x.num"│
╞═══════╡
│1      │
├───────┤
│2      │
├───────┤
│0      │
└───────┘
```
wrong order

## HINT
adding `CYPHER runtime=slotted` to the query fixes the issue

**Comment by henriknyman** [comment]
@matiasmorant Thanks for reporting!

Cypher does not actually guarantee the order in this case unless you explicitly put an `ORDER BY x.num` also on the final RETURN-clause, as explained here:

https://neo4j.com/docs/cypher-manual/current/clauses/order-by/#order-with
"The order is not guaranteed to be retained after the following clause, unless that also has an ORDER BY subclause."

But unfortunately there is indeed a bug here (even with an extra ORDER BY-clause) in that the Cypher planner thinks that the provided order is preserved through the subquery with the UNION, which is currently not true for `pipelined` runtime.

As you mention, prefixing the query with `runtime=slotted` is a suitable work-around, since it preserves the order through UNION.

**Comment by henriknyman** [comment]
@matiasmorant A fix was included in the Neo4j 4.2.10 release (changelog: https://github.com/neo4j/neo4j/wiki/Neo4j-4.2-changelog#4210). It is only a fix to make sure it gets the correct result however, and it could do so by planning an extra full sort, so it may be that in your particular case it may still be more efficient to run with `runtime=slotted`. In Neo4j 4.4 we are looking at the possibility of preserving the provided order more efficiently in `pipelined` runtime by using the `OrderedUnion` operator.
